account_payment_mode_auto_reconcile/tests/test_partner_auto_reconcile.py

```diff
@@ -0,0 +1,151 @@
+        cls.acc_rev = cls.env['account.account'].search(
+            [('user_type_id', '=', cls.env.ref(
+                'account.data_account_type_revenue').id
+              )], limit=1
+        )
+        cls.partner = cls.env['res.partner'].create({
+            'name': 'Test partner',
+            'customer': True,
+            'property_account_receivable_id': cls.acc_rec.id,
+            'property_account_payable_id': cls.acc_pay.id,
+        })
+        cls.payment_mode = cls.env.ref(
+            'account_payment_mode.payment_mode_inbound_dd1'
+        )
+        # TODO check why it's not set from demo data
+        cls.payment_mode.auto_reconcile_outstanding_credits = True
+        cls.product = cls.env.ref('product.consu_delivery_02')
+        cls.invoice = cls.env['account.invoice'].create({
+            'partner_id': cls.partner.id,
+            'type': 'out_invoice',
+            'payment_term_id': cls.env.ref('account.account_payment_term').id,
+            'account_id': cls.acc_rec.id,
+            'invoice_line_ids': [(0, 0, {
+                'product_id': cls.product.id,
+                'name': cls.product.name,
+                'price_unit': 1000.0,
+                'quantity': 1,
+                'account_id': cls.acc_rev.id,
+            })],
+        })
+        cls.invoice.action_invoice_open()
+        bank_journal = cls.env['account.journal'].search(
+            [('type', '=', 'bank')], limit=1
+        )
+        cls.invoice.pay_and_reconcile(bank_journal)
+        cls.refund_wiz = cls.env['account.invoice.refund'].with_context(
+            active_ids=cls.invoice.ids).create({
+                'filter_refund': 'refund',
+                'description': 'test'
+            })
+        refund_id = cls.refund_wiz.invoice_refund().get('domain')[1][2]
+        cls.refund = cls.env['account.invoice'].browse(refund_id)
+        cls.refund.action_invoice_open()
+        cls.invoice_copy = cls.invoice.copy()
+        cls.invoice_copy.write({
+            'invoice_line_ids': [(0, 0, {
+                'product_id': cls.product.id,
+                'name': cls.product.name,
+                'price_unit': 500.0,
+                'quantity': 1,
+                'account_id': cls.acc_rev.id,
+            })]
+        })
+        cls.invoice_copy.action_invoice_open()
+
+    def test_invoice_validate_auto_reconcile(self):
+        auto_rec_invoice = self.invoice.copy({
+            'payment_mode_id': self.payment_mode.id,
+        })
+        auto_rec_invoice.write({
+            'invoice_line_ids': [(0, 0, {
+                'product_id': self.product.id,
+                'name': self.product.name,
+                'price_unit': 500.0,
+                'quantity': 1,
+                'account_id': self.acc_rev.id,
+            })]
+        })
+        self.assertTrue(self.payment_mode.auto_reconcile_outstanding_credits)
+        self.assertEqual(self.invoice_copy.residual, 1500)
+        auto_rec_invoice.action_invoice_open()
```

Comment:
should this behavior run when we go from open -> done?

Reply:
I don't really understand your question. The auto reconciliation will be done when the invoice is open, and if there are enough outstanding credit to reconcile, the invoice will be set to done...

Reply:
if invoice wasn't fully reconciled and after new lines appeared, they will not be added here
if this action should performed only once than ok,
 maybe it can be done in action_invoice_validate and unreconcilation in action_invoice_cancel instead of write method

Reply:
As written in the description "Automatic reconciliation of outstanding credits will only happen on customer
invoices at validation if the payment mode is set or when the payment mode is
changed on an open invoice."